lambda_functions/two_way_sms/two_way_sms_v2.py

```diff
@@ -0,0 +1,369 @@
+    logger.setLevel("INFO")
+    logger.warning("Invalid log level specified.  Defaulting to INFO.")
+
+try:
+    TIMEOUT = tuple(TIMEOUT) if isinstance(TIMEOUT, list) else int(TIMEOUT)
+except (TypeError, ValueError):
+    TIMEOUT = 3
+    logger.warning("Invalid TIMEOUT value specified.  Defaulting to 3 seconds.")
+
+
+################################################################################################
+# Get the database URI from SSM Parameter Store.
+################################################################################################
+
+if os.getenv("NOTIFY_ENVIRONMENT") == "test":
+    sqlalchemy_database_uri = os.getenv("SQLALCHEMY_DATABASE_URI")
+else:
+    database_uri_path = os.getenv("DATABASE_URI_PATH")
+    if database_uri_path is None:
+        # Without this value, this code cannot know the path to the required
+        # SSM Parameter Store resource.
+        sys.exit("DATABASE_URI_PATH is not set.  Check the Lambda console.")
+
+    logger.debug("Getting the database URI from SSM Parameter Store . . .")
+    ssm_client = boto3.client("ssm", region_name=AWS_REGION)
+    ssm_response: dict = ssm_client.get_parameter(
+        Name=database_uri_path,
+        WithDecryption=True
+    )
+    logger.debug(". . . Retrieved the database URI from SSM Parameter Store.")
+    sqlalchemy_database_uri = ssm_response.get("Parameter", {}).get("Value")
+
+    if sqlalchemy_database_uri is None:
+        sys.exit("Can't get the database URI from SSM Parameter Store.")
+
+################################################################################################
+# Use the database URI to get a 10DLC-to-URL mapping from the database.
+# The format should be:
+#    {
+#        <phone number>: {
+#            'service_id': <value>,
+#            'url_endpoint': <value>,
+#            'self_managed': <value>,
+#        }
+#    }
+################################################################################################
+
+if os.getenv("NOTIFY_ENVIRONMENT") == "test":
+    two_way_sms_table_dict = {}
+else:
+    try:
+        logger.info("Retrieving the 10DLC-to-URL mapping from the database . . .")
+        db_connection = psycopg2.connect(sqlalchemy_database_uri)
+        if db_connection is None:
+            sys.exit("No database connection.")
+
+        with db_connection.cursor() as c:
+            # https://www.psycopg.org/docs/cursor.html#cursor.execute
+            c.execute(INBOUND_NUMBERS_QUERY)
+            mapping_data = c.fetchall()
+            logger.debug("Data returned from query: %s", mapping_data)
+
+        db_connection.close()
+        logger.info(". . . Retrieved the 10DLC-to-URL mapping from the database.")
+    except psycopg2.Warning as e:
+        logger.warning(e)
+    except psycopg2.OperationalError as e:        
+        logger.exception(e)
+        sys.exit("Unable to retrieve the 10DLC-to-URL mapping from the database.")
+    finally:
+        db_connection.close()
```

Comment:
```suggestion
    db_connection = None
    try:
        logger.info("Retrieving the 10DLC-to-URL mapping from the database . . .")
        db_connection = psycopg2.connect(sqlalchemy_database_uri)
        if db_connection is None:
            sys.exit("No database connection.")

        with db_connection.cursor() as c:
            # https://www.psycopg.org/docs/cursor.html#cursor.execute
            c.execute(INBOUND_NUMBERS_QUERY)
            mapping_data = c.fetchall()
            logger.debug("Data returned from query: %s", mapping_data)

        logger.info(". . . Retrieved the 10DLC-to-URL mapping from the database.")
    except psycopg2.Warning as e:
        logger.warning(e)
    except psycopg2.OperationalError as e:        
        logger.exception(e)
        sys.exit("Unable to retrieve the 10DLC-to-URL mapping from the database.")
    finally:
        if db_connection:
            db_connection.close()
```

IDK what happens if you try to close it twice. Also, if it threw an exception it may or may not have been set.

Reply:
Doh!  I copied when I meant to cut.

Reply:
https://www.psycopg.org/docs/connection.html#connection.closed